### DISCUSSION TITLE
Help Setting Up Padloc Self Host

### DISCUSSION THREAD

**Discussion by Skalhaxor** [padloc/padloc#412] [Q&A] [answered]
I've self hosted Padlock for a long time, and it worked great. But the recent update to the Android app that gives the deprecation notice has actually broken the sync to my self host. That's a whole other issue though, and not what this post is for.

With the deprecation of Padlock, I figured it was time to finally attempt to self host Padloc. I've been following the instructions here: https://padloc.app/help/guides/self-host/, but for the life of me, I cannot get it to work.

First of all, I can't even get the basic setup, without reverse proxy, to work. I copied the given configuration exactly, but I just get unable to connect in browser. Moving forward with the reverse proxy instructions, I am able to connect to Nginx using my hostname and SSL cert, but Nginx just reports 502 Bad Gateway. I even tried to setup logs for Nginx and the Padloc server, and Nginx says no route to host. But I am getting zero logs from Padloc. In fact, none of the padloc directories are being populated at all.

Are the self host instructions outdated and need to be updated? I noticed the environment variable defaults at https://github.com/padloc/padloc#configuration are slightly different from the ones in the self host setup. I'm running this on an arm64 Odroid C2 running Ubuntu Mate 20.04, but I'm not sure that matters since this is through Docker. Is there a target Docker version? My Docker engine version is 20.10.12, and Docker compose is 1.25.0.

**Comment by MaKleSoft** [comment]
Can you post the output of `docker-compose up` please?

**Comment by Skalhaxor** [comment, marked as answer]
I don't know why I never thought of that. I've just been putting it to the background with -d, so I never saw the output of the compose. Here is the output:
```
Creating network "padloc_default" with the default driver
Creating padloc_pwa    ... done
Creating nginx         ... done
Creating padloc_server ... done
Attaching to padloc_server, nginx, padloc_pwa
pwa_1     | standard_init_linux.go:228: exec user process caused: exec format error
server_1  | standard_init_linux.go:228: exec user process caused: exec format error
nginx     | /docker-entrypoint.sh: /docker-entrypoint.d/ is not empty, will attempt to perform configuration
nginx     | /docker-entrypoint.sh: Looking for shell scripts in /docker-entrypoint.d/
nginx     | /docker-entrypoint.sh: Launching /docker-entrypoint.d/10-listen-on-ipv6-by-default.sh
nginx     | 10-listen-on-ipv6-by-default.sh: info: Getting thechecksum of /etc/nginx/conf.d/default.conf
nginx     | 10-listen-on-ipv6-by-default.sh: info: Enabled listen on IPv6 in /etc/nginx/conf.d/default.conf
nginx     | /docker-entrypoint.sh: Launching /docker-entrypoint.d/20-envsubst-on-templates.sh
nginx     | /docker-entrypoint.sh: Launching /docker-entrypoint.d/30-tune-worker-processes.sh
nginx     | /docker-entrypoint.sh: Configuration complete; ready for start up
nginx     | 2022/03/04 11:48:31 [emerg] 1#1: host not found in upstream "padloc_server" in /etc/nginx/nginx.conf:25
nginx     | nginx: [emerg] host not found in upstream "padloc_server" in /etc/nginx/nginx.conf:25
padloc_server exited with code 1
padloc_pwa exited with code 1
nginx exited with code 1
```

Looks like both Padloc images are failing on `standard_init_linux.go:228: exec user process caused: exec format error`. Looking this up, I'm seeing a common cause is due to the Docker image being built for a different architecture than what it's run on. Looking at the Docker hub page for padloc/server, I'm only seeing amd64, but I'm running this on arm64.

Are the Padloc Docker images not built for arm64 then? If not, should I create an issue request to have arm64 built packages?

**Reply by MaKleSoft** [reply]
Yes, the images hosted on docker hub are amd only. I think the easiest solution would be for you to build the images from source. Download the source code, then update your `docker-compose.yml`:

```diff
 version: "3"
 services:
     server:
         image: padloc/server
+        build:
+            context: /path/to/padloc/source
+            dockerfile: Dockerfile-server
        ...
     pwa:
         image: padloc/pwa
+        build:
+            context: /path/to/padloc/source
+            dockerfile: Dockerfile-pwa
         ...
```

**Reply by Skalhaxor** [reply]
Thanks for the information! I finally got Padloc up and running on my server! I did have a couple additional things I needed to change in order to get it running, though.

I made the updates to my `docker-compose.yml` that you suggested above, but it was still trying to use the images pulled from Docker Hub. So I had to remove the `image` directives:
```diff
 version: "3"
 services:
     server:
-        image: padloc/server
+        build:
+            context: /path/to/padloc/source
+            dockerfile: Dockerfile-server
        ...
     pwa:
-        image: padloc/pwa
+        build:
+            context: /path/to/padloc/source
+            dockerfile: Dockerfile-pwa
         ...
```

Then when docker-compose was building from the source, I ran into this error:
```
Step 16/26 : RUN npm ci --unsafe-perm
 ---> Running in ff874f7b0c07
npm WARN old lockfile 
npm WARN old lockfile The package-lock.json file was created with an old version of npm,
npm WARN old lockfile so supplemental metadata must be fetched from the registry.
npm WARN old lockfile 
npm WARN old lockfile This is a one-time fix-up, please be patient...
npm WARN old lockfile 
npm notice 
npm notice New patch version of npm available! 8.5.1 -> 8.5.3
npm notice Changelog: <https://github.com/npm/cli/releases/tag/v8.5.3>
npm notice Run `npm install -g npm@8.5.3` to update!
npm notice 
npm ERR! `npm ci` can only install packages when your package.json and package-lock.json or npm-shrinkwrap.json are in sync. Please update your lock file with `npm install` before continuing.
npm ERR!
```

Docker was pulling the latest version of the node image, which looks like it no longer works with the Padloc v3.1.3 source I downloaded. I looked up the node version that was out around the time of the v3.1.3 release, which is v14. I modified both the `Dockerfile-server` and `Dockerfile-pwa` in the Padloc source to use v14 of node instead of the latest:
```diff
-FROM node
+FROM node:14
```
